source/game/system/KPadController.cc

```diff
@@ -147,6 +147,59 @@ void RaceInputState::reset() {
     trickRaw = 0;
 }
 
+/// @brief Checks if the input state is valid.
+/// @return If the input state is valid.
+bool RaceInputState::isValid() const {
+    // Check if there are any invalid buttons
+    // We cannot validate from the previous input state because it doesn't exist here
+    if (!!(buttons & ~0xf)) {
+        return false;
+    }
+
+    // Check if the sticks are in the domain
+    // The set of valid stick values is {(n - 7) / 7 | 0 <= n <= 14, in Z}
+    // It's possible for the stick input to be 8/7 with n = 15, but only with ghost controllers
+    if (stick.x > 1.0f || stick.y > 1.0f) {
+        return false;
+    }
+
+    for (size_t i = 0; i <= 14; ++i) {
+        auto xCond = stick.x <=> (static_cast<f32>(i) - 7.0f) / 7.0f;
+        ASSERT(xCond != std::partial_ordering::unordered);
+
+        if (xCond == std::partial_ordering::equivalent) {
+            break;
```

Comment:
I feel like each part of the validation should be in a separate helper function. It feels weird that in these loops and the switch below that `break` really means `true` meaning it's valid. Just calling helper functions from here would help better split apart the different segments of validation too.

Reply:
Agreed. Fixed.